### PR TITLE
Get rid of gcc-cross-wrapper

### PR DESCRIPTION
**Pull request by Ericson2314**
###### Motivation for this change

We want to get rid of the old gcc-cross-wrapper cause it is old, jank, and redundant. Also, it won't generalize to non-gcc for e.g cross compiling on Darwin. This PR just worries about linux, but ensuring no regressions there is work enough. Darwin support will be a done in a follow-up PR (probably repurposed #25047).

This looks stupendously big, but only the first few commits (already split for easier reviewing) actually change infrastructure. The rest are rote fix-ups to get tests to pass that do not need to be reviewed so closely.

###### Things done

http://hydra.nixos.org/jobset/nixpkgs/ericson2314-cross-scratch 2/3 of the PR is cleaning up derivations so those tests don't regress!

No native hashes should be changed by this commit

---

Huge shout-out to @Dridus for helping author this massive changeset!

CC @matthewbauer

### CODE REVIEW

File: pkgs/build-support/cc-wrapper/default.nix [hunk 1]
```diff
@@ -10,6 +10,10 @@
 , zlib ? null, extraPackages ? [], extraBuildCommands ? ""
 , dyld ? null # TODO: should this be a setup-hook on dyld?
 , isGNU ? false, isClang ? cc.isClang or false, gnugrep ? null
+, # Prefix for binaries. Customarily ends with a dash separator.
+  prefix ? stdenv.lib.optionalString (targetPlatform != hostPlatform)
+                                     (targetPlatform.config + "-")
```
Comment (matthewbauer) [inline pkgs/build-support/cc-wrapper/default.nix]: Should this be an argument for cc-wrapper? If you want a different prefix shouldn't you change the targetPlatform?

File: pkgs/build-support/cc-wrapper/default.nix [hunk 2]
```diff
@@ -162,87 +166,87 @@ stdenv.mkDerivation {
...
-        wrap gcc ${./cc-wrapper.sh} $ccPath/gcc
-        ln -s gcc $out/bin/cc
-        export real_cc=gcc
-        export real_cxx=g++
+      if [ -e $ccPath/${prefix}gcc ]; then
+        wrap ${prefix}gcc ${./cc-wrapper.sh} $ccPath/${prefix}gcc
+        ln -s ${prefix}gcc $out/bin/${prefix}cc
+        export real_cc=${prefix}gcc
+        export real_cxx=${prefix}g++
       elif [ -e $ccPath/clang ]; then
```
Comment (matthewbauer) [inline pkgs/build-support/cc-wrapper/default.nix]: If I follow correctly, this should be ```${prefix}clang```.

File: pkgs/build-support/cc-wrapper/default.nix [hunk 3]
```diff
@@ -162,87 +166,87 @@ stdenv.mkDerivation {
...
         export real_cxx=clang++
       fi
 
-      if [ -e $ccPath/g++ ]; then
-        wrap g++ ${./cc-wrapper.sh} $ccPath/g++
-        ln -s g++ $out/bin/c++
+      if [ -e $ccPath/${prefix}g++ ]; then
+        wrap ${prefix}g++ ${./cc-wrapper.sh} $ccPath/${prefix}g++
+        ln -s ${prefix}g++ $out/bin/${prefix}c++
       elif [ -e $ccPath/clang++ ]; then
```
Comment (matthewbauer) [inline pkgs/build-support/cc-wrapper/default.nix]: here too? unless something is special about clang

File: pkgs/build-support/cc-wrapper/default.nix [hunk 4]
```diff
@@ -162,87 +166,87 @@ stdenv.mkDerivation {
...
-        wrap clang++ ${./cc-wrapper.sh} $ccPath/clang++
-        ln -s clang++ $out/bin/c++
+        wrap ${prefix}clang++ ${./cc-wrapper.sh} $ccPath/clang++
+        ln -s ${prefix}clang++ $out/bin/${prefix}c++
```
Comment (matthewbauer) [inline pkgs/build-support/cc-wrapper/default.nix]: Ok so I'm not sure if I like this whole prefix thing. We generally just assume that /bin/cc and /bin/c++ are the *compilers I want to use*. Why not just leave it as is, and specify the compiler at the derivation level? The prefix thing just seems redundant when applied to Nix. I know that historically "gcc/cc" has referred to the "native" compiler, but in the Nix context, it's the **stdenv* * compiler right? Maybe I am just misunderstanding the purpose of prefixes but I'm just thinking of how annoying it would be to have to specify the right compiler I want to use for each mkDerivation call. For MacOS support, we do a "substituteInPlace --replace gcc cc" to get the clang compiler to be supported. The assumption is we will always have a compiler named "cc" that is available no matter what.

**Review by matthewbauer — CHANGES_REQUESTED** [review]

File: pkgs/build-support/cc-wrapper/default.nix (hunk 1 above)
Comment (Ericson2314) [inline pkgs/build-support/cc-wrapper/default.nix]: It used to not be a parameter, but I figured technically prefixing is a matter of taste, and users might want things to not be prefixed.

But it probably is a bad idea is the underlying compiler will already be prefixed or not, and as you noticed the llvm stuff isn't prefixed so it's conditional too.

So either I can just let-bind it, or I'll have a separate "in prefix" and "out prefix", and define is prefix to be non-`""` only when cross compiling *and* using GCC. Your call :).

File: pkgs/build-support/cc-wrapper/default.nix [hunk 5]
```diff
@@ -162,87 +166,87 @@ stdenv.mkDerivation {
...
       fi
 
-      export real_cc=cc
-      export real_cxx=c++
+      export real_cc=${prefix}cc
+      export real_cxx=${prefix}c++
       export default_cxx_stdlib_compile="${
-        if stdenv.isLinux && !(cc.isGNU or false)
-          then "-isystem $(echo -n ${cc.gcc}/include/c++/*) -isystem $(echo -n ${cc.gcc}/include/c++/*)/$(${cc.gcc}/bin/gcc -dumpmachine)"
+        if targetPlatform.isLinux && !(cc.isGNU or false)
```
Comment (Ericson2314) [inline pkgs/build-support/cc-wrapper/default.nix]: Ah! but this is not the `or` function (which also exists!) but the `or` operator providing a fallback if `isGNU` isn't present.

[N.B. this `or` works analogously to the `?` in a function pattern (`{ x ? foo }: bar`). I want to make an RFC so both are deprecated and `??` is used instead. `?` in an expression is unrelated and would stay the same.]

File: pkgs/build-support/cc-wrapper/default.nix (hunk 2 above)
Comment (Ericson2314) [inline pkgs/build-support/cc-wrapper/default.nix]: LLVM is multi-target---the unwrapped clang binaries support everything and thus don't need a prefix (which would require needless extra building).

Wrapped Clang however is target-specific as libc is target-specific, and even if libc is not part of cc-wrapper, we still want separate wrappers scrips so `NIX_CFLAGS_COMPILE` and friends don't get jumbled up (I sed those environment variables in a follow up commit).

File: pkgs/build-support/cc-wrapper/default.nix (hunk 4 above)
Comment (Ericson2314) [inline pkgs/build-support/cc-wrapper/default.nix]: Some derivations need to build stuff to be run as part of the build. This seems odd but seems to be quite common, actually. For those ones, we need to have both compilers on the PATH and not have them clobber.

In some grand future where we turn all makesfiles into Nix or something, hopefully we can do away with this, but until then we're out of luck.

File: pkgs/build-support/cc-wrapper/default.nix (hunk 1 above)
Comment (matthewbauer) [inline pkgs/build-support/cc-wrapper/default.nix]: Ok it's fine as is if you can just override it when you need to

**Review by matthewbauer — APPROVED** [review]
OK sounds good - at least it all makes some sense to me now!

File: pkgs/build-support/cc-wrapper/default.nix (hunk 1 above)
Comment (Ericson2314) [inline pkgs/build-support/cc-wrapper/default.nix]: I'll save the fancy stuff for later I think. let-bind for now!

**Comment by Ericson2314** [comment]
@Dridus Thanks again for agreeing to try shepherding this!

I've massively cleaned up the history here, and cherry-picked your zlib fix. Also I rewrote `dontSetConfigureCross` into something much nicer.

http://hydra.nixos.org/jobset/nixpkgs/ericson2314-cross-scratch is now set up to do all our cross tests for this PR. It should periodically pull the underlying branch every so often, and schedule builds accordingly. I can bump up the priority of the builds too. 

Basically, the goal is to adding commits fixing packages (well, I hope the prior commits are OK!) until there are no regressions.

**Comment by Ericson2314** [comment]
FYI eval errors are because `meta.platforms` is now being checked against the host platform.

**Comment by Ericson2314** [comment]
So I hydra has all cross builds passing but the build != Host == target GCCs. But I have a feeling they were broken anyways. Could anyone test, say, the aarch64 one from master and confirm?

CC @grahamc :)

**Comment by grahamc** [comment]
Tested by applying this diff:

```patch
diff --git a/pkgs/stdenv/linux/bootstrap-files/aarch64.nix b/pkgs/stdenv/linux/bootstrap-files/aarch64.nix
index ca7e84502a..99d45d7f81 100644
--- a/pkgs/stdenv/linux/bootstrap-files/aarch64.nix
+++ b/pkgs/stdenv/linux/bootstrap-files/aarch64.nix
@@ -4,8 +4,6 @@
     sha256 = "12qcml1l67skpjhfjwy7gr10nc86gqcwjmz9ggp7knss8gq8pv7f";
     executable = true;
   };
-  bootstrapTools = import <nix/fetchurl.nix> {
-    url = http://nixos-arm.dezgeg.me/bootstrap-aarch64-2017-03-11-bb3ef8/bootstrap-tools.tar.xz;
-    sha256 = "1075d5n4yclbhgisi6ba50601mw3fhivlkjs462qlnq8hh0xc7nq";
-  };
+  bootstrapTools = /nix/store/lwg73g7mxsbpv1nr2xhaam61b42rl596-stdenv-bootstrap-tools-cross-aarch64-unknown-linux-gnu/on-server/bootstrap-tools.tar.xz;
+  # bootstrapTools = "/nix/store/lwg73g7mxsbpv1nr2xhaam61b42rl596-stdenv-bootstrap-tools-cross-aarch64-unknown-linux-gnu/on-server/bootstrap-tools.tar.xz";
 }
```

and running `nix-build . -A hello` worked great:

```
$ file result/bin/hello 
result/bin/hello: ELF 64-bit LSB executable, ARM aarch64, version 1 (SYSV), dynamically linked, interpreter /nix/store/9br8761qwvvg7m301g47xz7yn9vm0ci7-glibc-2.25/lib/ld-linux-aarch64.so.1, for GNU/Linux 3.7.0, not stripped

$ ./result/bin/hello 
Hello, world!
```

**Comment by Ericson2314** [comment]
To be clear, that is *without* this PR. So the `gcc build != host == target` builds failures are indeed a regression.

File: pkgs/build-support/cc-wrapper/default.nix [hunk 6]
```diff
@@ -5,44 +5,109 @@
...
+
+    '' + stdenv.lib.optionalString (textFile == ./setup-hook.sh) ''
+      cat << 'EOF' >> $out
+        for CMD in ar as nm objcopy ranlib strip strings size ld
+        do
+          # which is not part of stdenv, but compgen will do for now
+          if
+            PATH=$_PATH type -p ${prefix}$CMD > /dev/null
+          then
+            export ''${CMD^^}=${prefix}''${CMD};
```
Comment (Dridus) [inline pkgs/build-support/cc-wrapper/default.nix]: this substitution doesn't seem to work on non-bash and older bashes (3.2)

Comment (Ericson2314) [inline pkgs/build-support/cc-wrapper/default.nix]: Do we use those? I'm fine changing it to is awk or whatever.

Comment (LnL7) [inline pkgs/build-support/cc-wrapper/default.nix]: Both bash and `stdenv.shell` are 4.4 now AFAIK.

Comment (Dridus) [inline pkgs/build-support/cc-wrapper/default.nix]: sorry, should have added context... it's true that in the build chain it'll be okay because all nix-controlled bash versions default to 4.4, but when using nix-shell it'll by default use the system bash, which on at least macOS 10.11 by default is 3.2. I can work around it with NIX_BUILD_SHELL=, but it makes for pretty awkward use of nix-shell when using the cross env.

Comment (dtzWill) [inline pkgs/build-support/cc-wrapper/default.nix]: The wrappers' shebang should point to an appropriate bash to ensure things work as expected, I'd think.  Quick check says my local cc-wrapper's are this way, is that not happening on Darwin?

Comment (Dridus) [inline pkgs/build-support/cc-wrapper/default.nix]: this is in the setup hook for the wrapper, not the wrapper proper. I imagine it wouldn't use the shebang anyway since it should be sourced into the enclosing shell environment? in any case, it's running in bash-3.2 on darwin

Comment (dtzWill) [inline pkgs/build-support/cc-wrapper/default.nix]: ah, okay. awk it is, then! (or whichever)

Comment (Dridus) [inline pkgs/build-support/cc-wrapper/default.nix]: I have a patch with `export $(echo "$CMD" | tr "[:lower:]" "[:upper:]")=${prefix}''${CMD}` that I'm waiting for a rebuild to confirm works correctly. this is on account of my non-knowledge of awk 😅

**Comment by Ericson2314** [comment]
Lynx is failing because something requires wants dynamic linking?
```
configure:19231: checking for tgoto in -lncurses
configure:19258: armv6l-unknown-linux-gnueabihf-cc -o conftest -O2  -D_FILE_OFFSET_BITS=64  -D_GNU_SOURCE  -DLINUX  -I/nix/store/hhl0iwwy4r0c8iyg0xj0qslqgqzvd0g8-openssl-1.0.2l-armv6l-unknown-linux-gnueabihf-dev/include  conftest.c -lncurses  -L/nix/store/2fnl9xk81540zvpq4s436c9iihqbdixv-openssl-1.0.2l-armv6l-unknown-linux-gnueabihf/lib -lssl -lcrypto >&5
/nix/store/l28dg0fy9553i4cg9g41zs3f2wwcjnrb-armv6l-unknown-linux-gnueabihf-binutils-2.28/bin/armv6l-unknown-linux-gnueabihf-ld: warning: libdl.so.2, needed by /nix/store/2fnl9xk81540zvpq4s436c9iihqbdixv-openssl-1.0.2l-armv6l-unknown-linux-gnueabihf/lib/libssl.so, not found (try using -rpath or -rpath-link)
/nix/store/2fnl9xk81540zvpq4s436c9iihqbdixv-openssl-1.0.2l-armv6l-unknown-linux-gnueabihf/lib/libcrypto.so: undefined reference to `dladdr@GLIBC_2.4'
/nix/store/2fnl9xk81540zvpq4s436c9iihqbdixv-openssl-1.0.2l-armv6l-unknown-linux-gnueabihf/lib/libcrypto.so: undefined reference to `dlsym@GLIBC_2.4'
/nix/store/2fnl9xk81540zvpq4s436c9iihqbdixv-openssl-1.0.2l-armv6l-unknown-linux-gnueabihf/lib/libcrypto.so: undefined reference to `dlerror@GLIBC_2.4'
/nix/store/2fnl9xk81540zvpq4s436c9iihqbdixv-openssl-1.0.2l-armv6l-unknown-linux-gnueabihf/lib/libcrypto.so: undefined reference to `dlopen@GLIBC_2.4'
/nix/store/2fnl9xk81540zvpq4s436c9iihqbdixv-openssl-1.0.2l-armv6l-unknown-linux-gnueabihf/lib/libcrypto.so: undefined reference to `dlclose@GLIBC_2.4'
collect2: error: ld returned 1 exit status
configure:19261: $? = 1
configure: failed program was:
#line 19239 "configure"
#include "confdefs.h"

/* Override any gcc2 internal prototype to avoid an error.  */
#ifdef __cplusplus
extern "C"
#endif
/* We use char because int might match the return type of a gcc2
   builtin and then its argument prototype would still apply.  */
char tgoto ();
int
main ()
{
tgoto ();
  ;
  return 0;
}
configure:19278: result: no
```

**Comment by Ericson2314** [comment]
Ok, I'm starting to think while there definitely is some regressions, the PR just is just getting bigger and older and it would be nice to merge and then deal with them---not many people are doing cross compilation anyways.

The known regressions are currently:

 - gcc `build != target == host`, only used when bootstrapping a new platform
 - mingw_w64, obsoleted by @DavidEGrayson's work long-term anyways
 - lynx on Raspberry Pi, just a random test to my knowledge

If we choose to deal with those later, the remain work is:
 - [ ] Use `tr` for old bash support (about to rebase for that)
 - [ ] Get a review as this is big

It's been a while since I had one of these  reviewed. Would @bjornfor @dezgeg be interested? @dezgeg I also give you final say on whether it's OK to temporarily break gcc like that.

**Comment by bjornfor** [comment]
I looked at all the commits in this PR just now. Big +1 from me. (Awesome job!)

Disclaimer: I didn't test anything. I don't cross compile with Nix (yet).

**Comment by Ericson2314** [comment]
@bjornfor thanks! And don't worry so much about actually building things. Hydra's got us covered for that.

**Comment by Ericson2314** [comment]
Fixed the darwin mass rebuild and applied @Dridus's `tr` patch.

**Comment by Ericson2314** [comment]
@dezgeg Any thoughts on the gcc regression? I'm very tempted to merge tomorrow so I can get on with the follow up.

File: pkgs/build-support/cc-wrapper/default.nix [hunk 7]
```diff
@@ -45,6 +45,12 @@ let
 
   default_cxx_stdlib_compile=optionalString (targetPlatform.isLinux && !(cc.isGNU or false))
     "-isystem $(echo -n ${cc.gcc}/include/c++/*) -isystem $(echo -n ${cc.gcc}/include/c++/*)/$(${cc.gcc}/bin/gcc -dumpmachine)";
+
+  dashlessTarget = stdenv.lib.replaceStrings ["-"] ["_"] targetPlatform.config;
```
Comment (matthewbauer) [inline pkgs/build-support/cc-wrapper/default.nix]: It looks like this kind of breaks with crossOverlays where targetPlatform == hostPlatform, but we want to only pass certain variables to cc-wrapper. For instance, this is used on macOS for semi-static binaries:

```
  # Best effort static binaries. Will still be linked to libSystem,
  # but more portable than Nix store binaries.
  makeStaticDarwin = stdenv: stdenv // {
    mkDerivation = args: stdenv.mkDerivation (args // {
      NIX_CFLAGS_LINK = toString (args.NIX_CFLAGS_LINK or "")
                      + " -static-libgcc";
      nativeBuildInputs = (args.nativeBuildInputs or []) ++ [ (makeSetupHook {
        substitutions = {
          libsystem = "${stdenv.cc.libc}/lib/libSystem.B.dylib";
        };
      } ../stdenv/darwin/portable-libsystem.sh) ];
    });
  };
```

NIX_CFLAGS_LINK should only be passed to nativeBuildInputs compilers. But, it is also being passed to depsBuildBuild compilers which are not guaranteed to be gcc. Can we choose another value for the salt? Not sure how to do this but it seems like every "cc-wrapper" should have a unique salt. Could we use the output path's hash?

Comment (matthewbauer) [reply pkgs/build-support/cc-wrapper/default.nix]: Avoiding the salt sounds good! This ends up kind of polluting the environment anyway.

Comment (Ericson2314) [reply pkgs/build-support/cc-wrapper/default.nix]: Note to self: env hook roles prevent getting rid of salt. Either we make it so the wrapper scripts consume no variables, or we make propagation logic happen on the nix level.

Comment (matthewbauer) [reply pkgs/build-support/cc-wrapper/default.nix]: This is mitigated by the fact that we can use llvm for cross now. I think we could just the same compiler for cross compilation as we do in stdenv.
